# Dawson Pallotta

# Update filename and headers

Answer questions in markdown cells. Please verify that your answers are formatted such that they are easy to read.

In [1]:
# imports all the packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.decomposition import PCA
from sklearn.utils import resample
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.metrics import confusion_matrix
from sklearn import tree
import xgboost as xgb
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeRegressor
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.metrics import mean_absolute_error
from statsmodels.tsa.arima.model import ARIMA

In [2]:
#imports the dataset for the notebook
df = pd.read_csv("C:/Users/student/Data Minning/ts.csv") 
print(df.shape)
df.head(2)

(369, 15)


id  day1  day2  day3  day4  day5  day6  day7  day8  day9  day10  \
0  1MZ0A5PA6    19     9     2     3    17    15    16     3    27     21   
1  E7EEQIEOV    12    14    15    25    21     5    25    17    14     10   

   day11  day12  day13  day14  
0     19     15     27      5  
1     14      9      7     14

## Simple Average 

In [3]:
#prints the mean absolute error for simple average
print("Mean absolute error: %.2f" % mean_absolute_error(df.day2, df.day1))

Mean absolute error: 6.05


What is the name of the above forecasting method? Add a header to this portion of the notebook accordingly.
the above forecasting method is simple average 
Explain what the MAE means in the domain context.
MAE is the average of the absolute error values 

## Moving and Weighted Moving Average 

In [4]:
#prints the mean absolute error for moving average and weighted moving average 
df2 = df[['day2', 'day3', 'day4', 'day5', 'day6', 'day7', 'day8', 'day9', 'day10', 'day11', 'day12', 'day13', 'day14']]
df5 = df[['day2', 'day3', 'day4', 'day5']]
slist2 = []
slist5 = []
for i in range(0, df.shape[0]):
    slist2.append(df2[i:i+1].sum().sum()/13)
    slist5.append(df5[i:i+1].sum().sum()/4)
print("Mean absolute error: %.2f" % mean_absolute_error(slist2, df.day1))
print("Mean absolute error: %.2f" % mean_absolute_error(slist5, df.day1))

Mean absolute error: 5.42
Mean absolute error: 5.80


What are the name of the above forecasting methods. Add a header to this portion of the notebook accordingly.
moving average and weighted moving average 
Which of the two methods from the above code performed better?

## ARIMA

In [5]:
#prints the mean absolute error for ARIMA
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(0, 0, 0)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Mean absolute error: 5.42


### First-order Autoregressive

In [6]:
#prints the mean absolute error for first-order autoregressive
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(1, 0, 0)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarni

Mean absolute error: 5.83


C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


### Random Walk

In [7]:
#prints the mean absolute error for random walk
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(0, 1, 0)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarni

Mean absolute error: 7.54


### Differenced First-order autoregressive

In [8]:
#prints the mean absolute error for differenced first-order autoregressive
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(1, 1, 0)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarni

Mean absolute error: 7.18


C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


### Simple Exponential Smoothing

In [9]:
#prints the mean absolute error for simple exponential smoothing
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(0, 1, 1)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\tsa\statespace\sarimax.py:978: UserWarning: Non-invertible starting MA parameters found. Using zeros as starting parameters.
  warn('Non-invertible starting MA parameters found.'
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: Converg

Mean absolute error: 6.65


C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


In [10]:
#prints the mean absolute error for an arima with an order of 1,1,1
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(1, 1, 1)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\tsa\statespace\sarimax.py:978: UserWarning: Non-invertible starting MA parameters found. Using zeros as starting parameters.
  warn('Non-invertible starting MA parameters found.'
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\tsa\statespace\sarimax.py:966: UserWarning: Non-stationary starting autoregressive parameters found. Using zeros as starting parameters.
  warn('Non-stationary starting autoregressive parameters'
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\

Mean absolute error: 6.52


C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


In [12]:
#prints the mean absolute error for ARIMA with an order of 0,0,5
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(0, 0, 5)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\tsa\statespace\sarimax.py:866: UserWarning: Too few observations to estimate starting parameters for ARMA and trend. All parameters except for variances will be set to zeros.
  warn('Too few observations to estimate starting parameters%s.'
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaco

Mean absolute error: 7.26


C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


In [13]:
#prints the mean absolute error for ARIMA with an order of 5,0,0
predlist = []
for i in range(0, df2.shape[0]):
    mod = ARIMA(list(df2.iloc[i]), order=(5, 0, 0)).fit()
    fc, se, conf = mod.forecast(3, alpha=0.05)
    predlist.append(fc)
    
print("Mean absolute error: %.2f" % mean_absolute_error(predlist, df.day1))

C:\Users\student\anaconda3\lib\site-packages\statsmodels\tsa\statespace\sarimax.py:966: UserWarning: Non-stationary starting autoregressive parameters found. Using zeros as starting parameters.
  warn('Non-stationary starting autoregressive parameters'
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\model.py:566: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "
C:\Users\student\anaconda3\lib\site-packages\statsmodels\base\mode

Mean absolute error: 7.17


Ignore the red warnings. This just indicates that not all of the 369 time series can be well modeled through ARIMA.

Copy the above block of code for different ARIMA variations.

Try (p,d,q) = (0,0,0). What other model name shares the same MAE?
Moving average 
Run the model for first-order, random walk, differenced first-order, and simple exponential smoothing. List and compare their MAE values.
First-order = 5.83
Random Walk = 7.54
Fifferenced First Order = 7.18 
Simple Exponential Smoothing = 6.65 
Try a few other ARIMA variations. Can you find one with a lower MAE score than the other ARIMA models?
No

General Questions

Which forecasting model had the lowest MAE?
The regular ARIMA and moving average 
Is this what you expected? Explain your answer.
Yes 